src/coreclr/jit/redundantbranchopts.cpp

```diff
@@ -1083,6 +1083,61 @@ bool Compiler::optRedundantRelop(BasicBlock* const block)
         // We going to forward-sub a copy of candidateTree
         //
         assert(!sideEffect);
+
+        // Note the candidateTree's result may still be live.
+        //
+        // We should only duplicate candidateTree tree if we're sure the copy
+        // will produce the same result. In particular if the tree contains
+        // volatile reads, it may not evaluate the same.
+        //
+        // Volatile reads are modelled as side effects in VN but not in the IR.
```

Comment:
I assume checking for `GTF_ORDER_SIDEEFF` resulted in too many (avoidable) regressions?

Reply:
Good question, let me check -- doing that would be simpler.

Reply:
That seems to work out just as well, so will update.